crates/turbopack-ecmascript/src/chunk/source_map.rs

```diff
@@ -136,16 +137,17 @@ impl AssetReference for EcmascriptChunkSourceMapAssetReference {
         );
         if self.hot_module_replacement {
             // Expose a SourceMap for each chunk item when HMR is enabled
-            for item in &self.chunk.chunk_content().await?.module_factories {
-                source_maps.push(
-                    EcmascriptChunkEntrySourceMapAsset {
-                        chunk_path: path,
-                        code: item.code_vc,
-                        id: item.chunk_item.id(),
-                    }
-                    .keyed_cell(EcmascriptChunkEntrySourceMapAssetCellKey(item.chunk_item))
-                    .into(),
-                );
+            for chunk_items_chunk in self
+                .chunk
+                .chunk_content_result()
+                .await?
+                .chunk_items
+                .await?
+                .iter()
+            {
+                for item in chunk_items_chunk.await?.iter() {
+                    source_maps.push(EcmascriptChunkEntrySourceMapAssetVc::new(path, *item).into());
```

Comment:
I'm a little lost with how this helps to not invalidate. As far as I understand, the `id` is just `ValueToString` of the item, so always constant until the items path changes (which shouldn't happen). The `code_vc` just depends on the `item.content()` cell, so it shouldn't change either.

Is it the `factory().await?` read that's causing the invalidations? So by deferring that until the `EcmascriptChunkEntrySourceMapAsset::content()`, we avoid recreating the `EcmascriptChunkEntrySourceMapAssetVc` itself. And in this case it should be ok, because if the factory's read were to change, the source map would necessarily change.

Reply:
`self.chunk.chunk_content().await?` is the problematic thing. It's a snapshot of the chunk content, so it changes everytime a chunk item changes (as it includes the hash).